packages/engine/src/worker/runner/javascript/mod.rs

```diff
@@ -559,135 +559,292 @@ impl<'m> RunnerImpl<'m> {
+        // means an unspecified value. We read [1, 2, 3, 4] from the JS data by using `data.len` and
+        // then resize the buffer to `target_len`.
         let target_len = len.unwrap_or(data.len);
 
-        // TODO: We currently copy the buffers by calling `Buffer::from_slice_ref` because the
-        //   JavaScript representation of arrays does not match the Rust implementation. Try to
-        //   reduce copies where possible.
-        let mut builder = ArrayData::builder(field.data_type().clone());
-
-        match field.data_type() {
-            DataType::Boolean => {
-                // SAFETY: `data.buffer_ptrs[0]` is provided by arrow, the type is `u8`, and
-                //   `target_len` is carefully chosen
-                let values = unsafe { slice::from_raw_parts(data.buffer_ptrs[0], target_len) };
-
-                // Booleans are packed in arrow, `values` are already packed, so use them directly
-                let mut boolean_builder = BooleanBufferBuilder::new(target_len);
-                boolean_builder.append_packed_range(0..target_len, values);
-
-                builder = builder.add_buffer(boolean_builder.finish());
-            }
-            DataType::UInt16 => {
-                // SAFETY: `data.buffer_ptrs[0]` is provided by arrow, the type is `u16`, and
-                //   `target_len` is carefully chosen
-                let values =
-                    unsafe { slice::from_raw_parts(data.buffer_ptrs[0] as *const u16, target_len) };
-
-                builder = builder.add_buffer(Buffer::from_slice_ref(&values));
-            }
-            DataType::UInt32 => {
-                // SAFETY: `data.buffer_ptrs[0]` is provided by arrow, the type is `u32`, and
-                //   `target_len` is carefully chosen
-                let values =
-                    unsafe { slice::from_raw_parts(data.buffer_ptrs[0] as *const u32, target_len) };
-
-                builder = builder.add_buffer(Buffer::from_slice_ref(&values));
-            }
-            DataType::Float64 => {
-                // SAFETY: `data.buffer_ptrs[0]` is provided by arrow, the type is `f64`, and
-                //   `target_len` is carefully chosen
-                let values =
-                    unsafe { slice::from_raw_parts(data.buffer_ptrs[0] as *const f64, target_len) };
-
-                builder = builder.add_buffer(Buffer::from_slice_ref(&values));
-            }
+        let mut builder = ArrayData::builder(data_type.clone());
+
+        match data_type {
+            DataType::Boolean => unsafe {
+                // SAFETY: `data` is provided by arrow
+                debug_assert!(
+                    !data.buffer_ptrs[0].is_null(),
+                    "Required pointer for `Boolean` (`buffers[0]`) is null"
+                );
+                builder = builder.add_buffer(self.read_boolean_buffer(
+                    NonNull::new_unchecked(data.buffer_ptrs[0] as *mut u8),
```

Comment:
In principle, I don't think this needs to be `mut`, since we're only reading the JS buffers for flushing, not writing to them.

Reply:
`NonNull` contains a `*mut` pointer. Casting this is safe to do.